### ISSUE TITLE
Auto-Scrolling/Jumping to Previous Location Link

### ISSUE THREAD

**Issue by a-paxton** [jupyter/notebook#699]
I'm having a curious problem when using links to other cells in the notebook.  After clicking on a link to another cell in the notebook, scrolling and clicking on any of my other cells -- either markdown or code cells only -- will cause the browser to jump back to the link destination.  (The jumping also occasionally occurs when simply scrolling after clicking the link.)  If I select a cell before it jumps up to the link cell, I can get the browser to jump back to the cell I'm working on until I use the space, delete, or return keys, at which point it jumps back to the link cell.  Any suggestions or insight would be greatly appreciated!

Chrome: 46.0.2490.80 (64-bit) 
Jupyter version: 4.04
Running on: Python 2.7.9 | Anaconda 2.2.0 (x86_64) | (default, Dec 15 2014, 10:37:34) 
[GCC 4.2.1 (Apple Inc. build 5577)]
Kernel: R version 3.2.2 (2015-08-14)

**Comment by ellisonbg** [comment]
I can reproduce. You don't even have to click on a link. If you render a Markdown cell with a heading, then click on the paragraph link to the right of the heading, the URL of the page will update to point to that location on the page, and then you will see the scroll problems if you try to scroll away.

Marking for 4.2 as I am worried that this is related to some scrolling changes we have made recently.

**Comment by ellisonbg** [comment]
Sorry, meant marking as 4.1. We should look at this before release to make sure it isn't a *new* bug.

**Comment by Carreau** [comment]
this is much more annoying that that. 
It's actually happening every time a notification is updated. 

Example:
```
IPython.notification_area.widget('kernel').element.toggle('slow')
```

It seem to be that Chome scoll to anchor on repaint/reflow.  Not sure what we can do (except removing the even on scroll that will mitigate the issue). 

Chris might now what to do ?

CC @jhamrick

**Comment by jhamrick** [comment]
That's bizarre. The problem also seems to go away if you do something like saving after clicking on the anchor link. I wonder if it has something maybe to do with what element is focused?

**Comment by Carreau** [comment]
Additional note, 

It **is** a reflow problem as if you **resize** the windows the element will scroll back to the anchor. I looked at window.selection, and documetn.activeElement i'm not sure if we can get around

**Comment by Carreau** [comment]
@sccolbert might have encountered that in Phosphor and have an idea ?

**Comment by sccolbert** [comment]
I think the issue is that Chrome tries to make focused input fields visible on screen, whether they have visibility set to hidden or not. CodeMirror uses one of these hidden input fields to capture key/text input. My guess is that it's scrolling to make sure that hidden field is "visible" in the viewport.

**Comment by jhamrick** [comment]
Hmm, so I've narrowed it down to if you call `IPython.notebook.create_checkpoint()` then it causes the issue to go away, but `IPython.notebook.create_checkpont_success()` by itself doesn't, so that means that something that the contents API is doing causes the problem to go away. If I do something else with the contents API (like listing checkpoints), it also seems to cause the problem to go away. So maybe we can look at what's going on there and find a solution to the problem?

@sccolbert would that behavior be consistent with CodeMirror's hidden input fields? It sounds to me like those would be separate things, but maybe there's something under the hood I don't know about.

**Comment by Carreau** [comment]
I have an ugly fix: 
```
h.append(
                        $('<a/>')
                            .addClass('anchor-link')
                            .attr('href', '#' + hash)
                            .text('¶')
                            .on('click',function(){
                                setTimeout(function(){that.unrender(); that.render()}, 100)
                            })
                    );
```

**Comment by jhamrick** [comment]
Yeah, that seems to fix it for me, too.

**Comment by sccolbert** [comment]
@jhamrick if we have code mirror instances on the page, they are going to have their hidden input fields lying around. I don't have the time to dig into it to see if that's the actual cause, but that's where I would start looking if I did.

**Comment by jhamrick** [comment]
Another hacky fix which is maybe slightly less overhead is to use `$.ajax()` instead of unrendering and re-rendering the cell.

**Comment by Carreau** [comment]
> to use $.ajax()

to do what ? Just empty call works ?

**Comment by jhamrick** [comment]
Yeah. Don't ask my why, because I have no idea, but it seems to work for some reason.

**Comment by Carreau** [comment]
Also we might want to do that on all the  `html.find("a[href]").not('[href^="#"]')` as people can create their own link, though we need to do it on the target cell that have the link, not the source cell, which is more difficult.

**Comment by jhamrick** [comment]
Yeah, that's a good idea, though you also don't want to do it on all links, just those that are anchor links. I think it would be sufficient to do find the `a` tags like you have there but then filter through them to only apply the function to those that start with `#`.

**Comment by sccolbert** [comment]
I think we should figure out what the root cause is and fix that, rather than adding undocumented hacks all of the place.

**Comment by jhamrick** [comment]
I agree it would be better to figure out what the root cause is, but I unfortunately don't have more time to spend on tracking it down today.

**Comment by ellisonbg** [comment]
This is the type of nasty thing that will bite us in unexpected ways unless
we find the root cause and fix it properly. I think we should live with the
bug until we find the root cause.

On Thu, Nov 5, 2015 at 5:30 PM, Jessica B. Hamrick <notifications@github.com
> wrote:

> I agree it would be better to figure out what the root cause is, but I
> unfortunately don't have more time to spend on tracking it down today.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jupyter/notebook/issues/699#issuecomment-154216110>.
>



-- 
Brian E. Granger
Associate Professor of Physics and Data Science
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com

**Comment by flothesof** [comment]
Hello everyone. I'm having the problem described in this issue. Is there are way I can (even temporarily) fix this while you find ways around the root cause? (I really appreciate your work!)
Thanks for your help.

**Comment by jhamrick** [comment]
@flothesof #706 should work as a temporary fix for now.

**Comment by JanSchulz** [comment]
I also see this, but I also use a TOC plugin, so clicking a link to a headline is happening a lot with me :-(

**Comment by Carreau** [comment]
I'll reopen to think of a real fix, and not a workaround. Once we fix it we ca revert #706.

**Comment by JanSchulz** [comment]
The workaround (I manually patched textcell.js, cleaned cache, restarted the browser and the server and used private mode to open a notebook) also does not work when you use a ToC extension, which links to headlines. Interestingly, it behaves badly after I click on such a link (which changes the URL to `http://localhost:8888/notebooks/whatever.ipynb#headline-text`), but not when I (re)load such a tab (which also switches to that place). 

Another observation: after clicking a link, I can't click on anything without jumping back. Even when I try to click on another headline "P" (which displays the right new link on hoover), I jump back to original selected headline. In "Inspect element", I see that the right cells get the "selected/unselected" tag, but I end up with the originally selected cell showing on the top of the currently visible part of the page.

This makes the ToC extensions basically useless :-(

**Comment by jfbercher** [comment]
Just to point out here that the workaround above is *only* for the toc extension in the IPython-contrib repo, and is essentially a simple call $.ajax() as suggested above by @jhamrick. The root of the problem is still unclear.

**Comment by plumSemPy** [comment]
Guys the problem still exists, every time I save or when the notebook autosaves it jumps to my latest markdown heading. While coding, it is a real nuisance. This page is the only ray of hope offering a solution, but could you please elaborate a bit more on what and how and where to $.ajax()? as I have no idea at this point in my career. Thank you.

**Comment by Carreau** [comment]
> Guys the problem still exists, every time I save or when the notebook autosaves it jumps to my latest markdown heading. While coding, it is a real nuisance. This page is the only ray of hope offering a solution, but could you please elaborate a bit more on what and how and where to $.ajax()? as I have no idea at this point in my career. Thank you.

are you using the TOC extension ?  And which version of the notebook ? Because that should have been fixed by #699.

**Comment by plumSemPy** [comment]
I am not using a TOC, although I am very much tempted, but I am concerned it will make the jumping worse. I am using jupyter 4.1.0. I upgraded less than a month ago, but I just upgraded again and restarted my session, and so far it seems there are no jumps. I will report back of the persistency of the issue or lack of it. Thanks.

**Comment by Carreau** [comment]
> I upgraded less than a month ago, but I just upgraded again and restarted my session, and so far it seems there are no jumps

Ok great ! Sorry about the jump thing. 



> will report back of the persistency of the issue or lack of it. Thanks.

Great ! Appreciate it !

**Comment by plumSemPy** [comment]
Oh not at all, I highly appreciate the effort you put into jupyter. So far I did not have an issue anymore. I do not know what it was, because I updated less than a month ago and I updated again and it seemed to have gone away. I will report if I encounter it in future, thanks.
